src/header-generator.js

```diff
@@ -0,0 +1,310 @@
+ *  in the language format accepted by that header, for example `en`, `en-US` or `de`.
+ * @param {string} httpVersion - Http version to be used to generate headers (the headers differ depending on the version).
+ *  Can be either 1 or 2. Default value is 2.
+ */
+
+/**
+ * HeaderGenerator randomly generates realistic browser headers based on specified options.
+ */
+class HeaderGenerator {
+    /**
+     * @param {HeaderGeneratorOptions} options - default header generation options used unless overridden
+     */
+    constructor(options = {}) {
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        this.defaultOptions = options;
+        this.headersOrder = JSON.parse(fs.readFileSync(headersOrderFilePath, { encoding: 'utf8' }));
+        const uniqueBrowserStrings = JSON.parse(fs.readFileSync(browserHelperFilePath, { encoding: 'utf8' }));
+        this.uniqueBrowsers = [];
+        for (const browserString of uniqueBrowserStrings) {
+            if (browserString === missingValueDatasetToken) {
+                this.uniqueBrowsers.push({
+                    name: missingValueDatasetToken,
+                });
+            } else {
+                this.uniqueBrowsers.push(prepareHttpBrowserObject(browserString));
+            }
+        }
+        this.inputGeneratorNetwork = new BayesianNetwork(inputNetworkDefinitionPath);
+        this.headerGeneratorNetwork = new BayesianNetwork(headerNetworkDefinitionPath);
+    }
+
+    /**
+     * Generates a single set of headers using a combination of the default options specified in the constructor
+     * and their possible overrides provided here.
+     * @param {HeaderGeneratorOptions} options - specifies options that should be overridden for this one call
+     */
+    getHeaders(options = {}) {
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        const headerOptions = { ...this.defaultOptions, ...options };
+
+        // Set up defaults
+        if (!headerOptions.locales) {
+            headerOptions.locales = ['en-US'];
+        }
+        if (!headerOptions.httpVersion) {
+            headerOptions.httpVersion = '2';
+        }
+        if (!headerOptions.browsers) {
+            headerOptions.browsers = [
+                { name: 'chrome' },
+                { name: 'firefox' },
+                { name: 'safari' },
+            ];
+        }
+        if (!headerOptions.operatingSystems) {
+            headerOptions.operatingSystems = [
+                'windows',
+                'macos',
+                'linux',
+                'android',
+                'ios',
+            ];
+        }
+
+        headerOptions.browsers = headerOptions.browsers.map((browserObject) => {
+            if (!browserObject.httpVersion) {
+                browserObject.httpVersion = headerOptions.httpVersion;
+            }
```

Comment:
Correct me if I'm wrong, but:

- If the user does not provide `headerOptions.httpVersion` it will be set to `2`.
- If the user does not provide `headerOptions.browsers` or provides them, but without `browserObject.httpVersion`, the `browserObject.httpVersion` will be set to `headerOptions.httpVersion` which will be `2`. 

Both of these situations will happen quite often I assume. Now, the user did not specify a browser version, yet, I assume, for browsers that we don't have any HTTP2 headers for, this will throw and tell the user to "relax" the requirements. The user can't relax anything though, because the problem is caused by the defaults.

Reply:
That is precisely what would happen. As to the "they can't relax anything" point, another thing the error mentions is to change something in the requirements, which is what they would need to do here. Anyway, this is a logical consequence of enforcing the default of 2. If you think that it shouldn't work like this, maybe we can prefer HTTP 2 if the user doesn't specify he wants HTTP 1, but allow the generator to generate HTTP 1 headers if no HTTP 2 headers can be generated. Would that be better for the users? What does @petrpatek think?

Reply:
I think this is a theoretical problem. I think we support only browsers we have data for - `chrome`, `firefox`, `safari` for http2. So the proposed scenario should never happen. If this happens, we would lack the support of the particular browser. 

We cant substitute HTTP2 headers with HTTP1 headers since it would fail the request because of incompatible headers such as `Connection`. 

If the user would specify more than one browser, let's say FF and safari. Safari would lack HTTP2 headers. The header generator would still produce HTTP2 headers for FF and ignore safari, am I right? Following this logic, the relaxing of requirements, in this case, means changing the browser, because there are no data for it. Once again this should never happen with default HTPP2.

Reply:
It can happen for some browser versions, I guess - if someone picks a version of a browser (and only that one version) old enough that there are no HTTP 2 items in our data for that browser version.
You are right about the two-browsers case.
Anyway, @petrpatek and @mnmkng, please, discuss how this should work and let me know what you decide is best.

Reply:
If @petrpatek says it's ok then it's ok. I just wanted to know that this is not something that slipped through the cracks.